core/src/test/java/org/bitcoinj/core/WalletTest.java

```diff
@@ -851,6 +852,418 @@ public void onWalletChanged(Wallet wallet) {
+            wallet.addTransactionsDependingOn(txns, wallet.getTransactions(true));
+            assertEquals(3, txns.size());
+            assertTrue(txns.contains(send1));
+            assertTrue(txns.contains(send1b));
+            assertTrue(txns.contains(send1c));
+        } finally {
+            wallet.setCoinSelector(originalCoinSelector);            
+        }
+        
+    }
+    
+    @Test
+    public void sortTxnsByDependency() throws Exception {
+        CoinSelector originalCoinSelector = wallet.getCoinSelector();        
+        try {
+            wallet.setCoinSelector(AllowUnconfirmedCoinSelector.get());
+            Transaction send1 = sendMoneyToWallet(valueOf(2, 0), AbstractBlockChain.NewBlockType.BEST_CHAIN);
+            final Address address = new ECKey().toAddress(params);
+            Transaction send1a = checkNotNull(wallet.createSend(address, valueOf(1, 0)));
+            wallet.commitTx(send1a);
+            Transaction send1b = checkNotNull(wallet.createSend(address, valueOf(0, 50)));
+            wallet.commitTx(send1b);
+            Transaction send1c = checkNotNull(wallet.createSend(address, valueOf(0, 25)));
+            wallet.commitTx(send1c);
+            Transaction send1d = checkNotNull(wallet.createSend(address, valueOf(0, 12)));
+            wallet.commitTx(send1d);
+            Transaction send1e = checkNotNull(wallet.createSend(address, valueOf(0, 06)));
+            wallet.commitTx(send1e);
+
+            Transaction send2 = sendMoneyToWallet(valueOf(200, 0), AbstractBlockChain.NewBlockType.BEST_CHAIN);
+            
+            SendRequest req2a = SendRequest.to(address, valueOf(100, 0));
+            req2a.tx.addInput(send2.getOutput(0));
+            req2a.shuffleOutputs = false;
+            wallet.completeTx(req2a);
+            Transaction send2a = req2a.tx;
+
+            SendRequest req2b = SendRequest.to(address, valueOf(50, 0));
+            req2b.tx.addInput(send2a.getOutput(1));
+            req2b.shuffleOutputs = false;
+            wallet.completeTx(req2b);
+            Transaction send2b = req2b.tx;
+
+            SendRequest req2c = SendRequest.to(address, valueOf(25, 0));
+            req2c.tx.addInput(send2b.getOutput(1));
+            req2c.shuffleOutputs = false;
+            wallet.completeTx(req2c);
+            Transaction send2c = req2c.tx;
+            
+            Set<Transaction> unsortedTxns = new HashSet<Transaction>();
+            unsortedTxns.add(send1a);
+            unsortedTxns.add(send1b);
+            unsortedTxns.add(send1c);
+            unsortedTxns.add(send1d);
+            unsortedTxns.add(send1e);
+            unsortedTxns.add(send2a);
+            unsortedTxns.add(send2b);
+            unsortedTxns.add(send2c);
+            List<Transaction> sortedTxns = wallet.sortTxnsByDependency(unsortedTxns);
+            
+            assertEquals(8, sortedTxns.size());
+            assertTrue(sortedTxns.indexOf(send1a) < sortedTxns.indexOf(send1b));
```

Comment:
Because the exact order of txns in sortedTxns is not guaranteed. Eg send1a,send1b,send1c,send1d,send1e,send2a,send2b,send2c is a valid result but send2a,send2b,send2c,send1a,send1b,send1c,send1d,send1e is also valid.